File: src/typescript/src/xiaoyi/domain/token/syntax/delimiter.ts
/**
 * # Delimiters
 *
 * `delimiter` defines all delimiter tokens (brackets, braces, parens, etc.).
 *
 * Path: `xiaoyi.domain.token.syntax.delimiter`
 *
 * - Layer 0: `domain`
 * - Layer 1: `token`
 * - Layer 2: `syntax`
 * - Layer 3: `delimiter`
 *
 * @module domain.token.syntax.delimiter
 * @brief Delimiter tokens (brackets, braces, parens)
 * @group Domain
 * @since 0.1.0
 * @author Miruamel
 * @see domain.token.syntax
 * @see domain.token.syntax.keyword
 */
import { SyntaxKind } from "..";

/** Delimiter token. */
export interface Delimiter {
  /** Opening character. */
  readonly open: string;
  /** Closing character. */
  readonly close: string;
  /** Delimiter kind. */
  readonly kind: DelimiterKind;
}

/** Delimiter categories. */
export enum DelimiterKind {
  /** Parentheses () */
  Paren = "paren",
  /** Square brackets [] */
  Bracket = "bracket",
  /** Curly braces {} */
  Brace = "brace",
  /** Angle brackets <> */
  Angle = "angle",
}

/** All delimiters. */
export const DELIMITERS: readonly Delimiter[] = Object.freeze([
  { open: "(", close: ")", kind: DelimiterKind.Paren },
  { open: "[", close: "]", kind: DelimiterKind.Bracket },
  { open: "{", close: "}", kind: DelimiterKind.Brace },
  { open: "<", close: ">", kind: DelimiterKind.Angle },
] as const);

/**
 * Get closing delimiter for opening.
 *
 * @param open - Opening character
 * @returns Closing character if valid delimiter
 * @since 0.1.0
 */
export function matchingClose(open: string): string | undefined {
  return DELIMITERS.find((d) => d.open === open)?.close;
}

/**
 * Get opening delimiter for closing.
 *
 * @param close - Closing character
 * @returns Opening character if valid delimiter
 * @since 0.1.0
 */
export function matchingOpen(close: string): string | undefined {
  return DELIMITERS.find((d) => d.close === close)?.open;
}

/**
 * Check if character is an opening delimiter.
 *
 * @param c - Character
 * @returns true if opening delimiter
 * @since 0.1.0
 */
export function isOpenDelimiter(c: string): boolean {
  return DELIMITERS.some((d) => d.open === c);
}

/**
 * Check if character is a closing delimiter.
 *
 * @param c - Character
 * @returns true if closing delimiter
 * @since 0.1.0
 */
export function isCloseDelimiter(c: string): boolean {
  return DELIMITERS.some((d) => d.close === c);
}

/**
 * Check if characters form a valid delimiter pair.
 *
 * @param open - Opening character
 * @param close - Closing character
 * @returns true if valid pair
 * @since 0.1.0
 */
export function isDelimiterPair(open: string, close: string): boolean {
  return matchingClose(open) === close;
}

File: src/python/xiaoyi/domain/token/syntax/delimiter.py
"""
# Delimiters

`delimiter` defines all delimiters (brackets, parentheses, etc.) in the Xiaoyi language.

Path: `xiaoyi.domain.token.syntax.delimiter`

- Layer 0: `domain`
- Layer 1: `token`
- Layer 2: `syntax`
- Layer 3: `delimiter`

@module xiaoyi.domain.token.syntax.delimiter
@brief Language delimiters
@group Domain
@since 0.1.0
@author Miruamel
@see xiaoyi.domain.token.syntax
@see xiaoyi.domain.token.syntax.operator
"""

from dataclasses import dataclass
from enum import Enum
from .kinds import SyntaxKind


@dataclass(frozen=True)
class Delimiter:
    """Delimiter token."""

    #: Delimiter text.
    text: str
    #: Delimiter kind.
    kind: "DelimiterKind"


class DelimiterKind(str, Enum):
    """Delimiter categories."""

    #: Opening delimiter.
    OPEN = "open"
    #: Closing delimiter.
    CLOSE = "close"


#: All delimiters.
DELIMITERS = tuple(
    Delimiter(text=t, kind=k)
    for t, k in [
        ("(", DelimiterKind.OPEN),
        (")", DelimiterKind.CLOSE),
        ("[", DelimiterKind.OPEN),
        ("]", DelimiterKind.CLOSE),
        ("{", DelimiterKind.OPEN),
        ("}", DelimiterKind.CLOSE),
        ("<", DelimiterKind.OPEN),
        (">", DelimiterKind.CLOSE),
    ]
)


def matching_close(open_delim: str) -> str | None:
    """
    Get matching closing delimiter.

    @param open_delim Opening delimiter
    @return Closing delimiter or None
    @since 0.1.0
    """
    pairs = {"(": ")", "[": "]", "{": "}", "<": ">"}
    return pairs.get(open_delim)


def matching_open(close_delim: str) -> str | None:
    """
    Get matching opening delimiter.

    @param close_delim Closing delimiter
    @return Opening delimiter or None
    @since 0.1.0
    """
    pairs = {")": "(", "]": "[", "}": "{", ">": "<"}
    return pairs.get(close_delim)


def is_open_delimiter(d: str) -> bool:
    """
    Check if string is an opening delimiter.

    @param d String to check
    @return True if opening delimiter
    @since 0.1.0
    """
    return d in {"(", "[", "{", "<"}


def is_close_delimiter(d: str) -> bool:
    """
    Check if string is a closing delimiter.

    @param d String to check
    @return True if closing delimiter
    @since 0.1.0
    """
    return d in {")", "]", "}", ">"}


def is_delimiter_pair(open_d: str, close_d: str) -> bool:
    """
    Check if two delimiters form a matching pair.

    @param open_d Opening delimiter
    @param close_d Closing delimiter
    @return True if matching pair
    @since 0.1.0
    """
    return matching_close(open_d) == close_d


__all__ = [
    "Delimiter",
    "DelimiterKind",
    "DELIMITERS",
    "matching_close",
    "matching_open",
    "is_open_delimiter",
    "is_close_delimiter",
    "is_delimiter_pair",
]